DecoratedTypeName.hpp.in

```diff
@@ -0,0 +1,40 @@
+#ifndef UTILITY_DECORATEDTYPENAME_HPP
+#define UTILITY_DECORATEDTYPENAME_HPP
+
+#include "utility.hpp"
+
+namespace utility{
+
+/**
+ * @brief 
+ * Provides the typename of a base class subject to compile-time a decorator 
+ *
+ * @details
+ * Acts as a base case terminating the template recursion of the more general
+ * variadic template. 
+ */
+template< typename Base, template< typename > class Decorator >
+struct DecoratedTypeName< Base, Decorator >{
+  using value = Decorator< Base >;
+};
+
+/**
+ * @brief 
+ * Provides the typename of a base class subject to compile-time decorators
+ *
+ * @details
+ * A recursive template evaluating the type name of a base class subject to
+ * two or more compile time decorator templates. Templates applied in right to 
+ * left order. 
+ */
+template< typename Base,
+          template< typename > class Decorator,
+          template< typename > class... RemainingDecorators >
+struct DecoratedTypeName< Base, Decorator, RemainingDecorators... >{
+  using value = Decorator< typename DecoratedTypeName
+                           < Base, RemainingDecorators... >::value >;
```

Comment:
This is neat stuff. This way we can extract the underlying type that was decorated?

If I take the example of the ChannelData for which we will have a base implementation for neutrons (called NeutronChannelData) and I use a decorator to replace for instance the penetrability and one of the channel radii. If I use this, I get the NeutronChannelData type out of it and none of the intermediate ones?

Is it possible to "de"-decorate each level separately? This way, I would be able to determine that for this particular ChannelData, the user has decided to override both the penetrability and channel radius. Such a feature would be usefull when writing the data to an ENDF or GND file.